### ISSUE TITLE
Result images with pre-trained models

### ISSUE THREAD

**Issue by Allycen** [OlavHN/fast-neural-style#1]
It seems that you have not overcome the problem which author of first chainer implementation of artistic style was faced even when switch on VGG19 instead of VGG16. Results are still very noisy. Could you explain the reason? Is it due to algorithm of generating pretraining models or something else? 
Just wanna generate images exactly the same as TensorFlow neural artistic style that you have mentioned

**Comment by OlavHN** [comment]
@jcjohnson: This is a reimplementation of your paper, "Perceptual Losses for Real-Time Style Transfer and Super-Resolution".

I seem to be getting a regular grid-like pattern in images generated by the transformation network. Another implementation seems to have gotten the same. See examples of what I mean: http://olavnymoen.com/2016/07/07/image-transformation-network .

In the paper all the images are too small to see if the same noise appeared in your implementation as well. **Could you confirm you got them all smooth or did you see a similar pattern?**

The grid like pattern makes me suspicious of the last 9x9 convolution, but I haven't experimented too much. The Total Variation loss does not seem to help much with the issue.

**Comment by jcjohnson** [comment]
I had some low-level noise, but not nearly as bad as the examples you linked. I've attached some of the raw images from Figure 6.

Here are baseline outputs for "Starry Night":

![28_baseline](https://cloud.githubusercontent.com/assets/2718714/16924404/02090f46-4cd4-11e6-8b7e-802f343a49cc.png)
![26_baseline](https://cloud.githubusercontent.com/assets/2718714/16924408/039ac624-4cd4-11e6-9467-b2ceb731e581.png)

Here are my outputs for "Starry Night":

![28_mine](https://cloud.githubusercontent.com/assets/2718714/16924421/0f236014-4cd4-11e6-8cde-a0e3196e7ee5.png)
![26_mine](https://cloud.githubusercontent.com/assets/2718714/16924423/108ef080-4cd4-11e6-8d40-95dc3a015d27.png)

Here are baseline outputs for "The Muse":

![21_baseline](https://cloud.githubusercontent.com/assets/2718714/16924435/1e762a42-4cd4-11e6-8fe2-2cc4e85fbfe8.png)
![31_baseline](https://cloud.githubusercontent.com/assets/2718714/16924436/1f718842-4cd4-11e6-8da5-dc59eceed341.png)

And here are my outputs:

![21_mine](https://cloud.githubusercontent.com/assets/2718714/16924438/240f34da-4cd4-11e6-830d-86c345bed846.png)
![31_mine](https://cloud.githubusercontent.com/assets/2718714/16924440/258d1d40-4cd4-11e6-8030-b62ffedb9370.png)

I didn't look through your implementation in detail so I'm not sure why you are getting such pronounced artifacts.

**Comment by hzy46** [comment]
@OlavHN Hello! Recently I have tried your code but the result is not very good. Then I experimented on one thing: only use the "style loss" to find the image to minimize it. (It is done by set CONTENT_WEIGHT=0 , TV_WEIGHT=0, and STYLE_WEIGHT=1 )

In the origin paper, for the picture "starry night", it should look like:
![qq 20161207174536](https://cloud.githubusercontent.com/assets/7499023/20962630/0bf5556e-bca5-11e6-9251-902274176e3b.jpg)

But this code generates:
![individualimage 1](https://cloud.githubusercontent.com/assets/7499023/20962603/eb42ab96-bca4-11e6-9df9-01965bda0b16.png)

I'm deciding to dive into the code more deeply. Could you give me some suggestions or hints on why this happens? Thank you.
(I'm concerned that it is because of the difference between VGG16 and VGG19.)

**Comment by jcjohnson** [comment]
Figure 4 in the paper was produced using optimization as implemented in slow_neural_style.lua. Your result looks like you used train.lua to train a feedforward model with no content weight; as you've discovered, this doesn't work very well.

**Comment by hzy46** [comment]
@jcjohnson Thank you very much for your response!

Just like what I said, when using the "transformation network" to minimize the "style loss", I found the generated image is filled with artifacts.

Then I train the whole model. The result picture is also filled with this kind of artifacts(like chessboard):

![res](https://cloud.githubusercontent.com/assets/7499023/20995526/abac60d8-bd32-11e6-8e98-51d7aea56009.jpg)

Maybe this implementation has some problem. I'm trying hard to use the "tv loss" to remove these artifacts.

**Comment by jcjohnson** [comment]
That result doesn't look very good to me; I'm not sure if it's a problem with hyperparameters or some bug in the implementation. You can try my implementation here:

https://github.com/jcjohnson/fast-neural-style

**Comment by OlavHN** [comment]
Hello @hzy46 !

As @jcjohnson is suggesting there is an error in my implementation of his paper somewhere.

If you'd like to dive in I'd be super happy to provide any help. A couple of things I've tried during (fruitless) debugging:

- I've modified the model to implement Johnsons "Super resolution" model, and that came out very well (very similar model without the style and tv loss)
- I've tried a lot with different padding schemes for the convolutions, and they make a huge difference in the amount of noise both at the edges and elsewhere. But no silver bullet.
- I've looked over the style loss calculation a bunch of times as I suspect here might be something there. So far no dice!
- Finally I implemented some of the tricks in http://distill.pub/2016/deconv-checkerboard/ , diverging from the paper. While that hides the issue somewhat, there is still some bug in there somewhere :)

Johnsons own implementation is now available so that is 1) a good alternative or 2) a great resource for tracking down my bugs, but I haven't spent time doing that.

**Comment by hzy46** [comment]
@OlavHN Thank you for the suggestions! 

I'm now carrying out more experiments.  If I can find the bug, I will post it in this issue. :-)

**Comment by hzy46** [comment]
I have done some experiments and here are my opinions:

I combine your code with Tensorflow Slim, so I can change from VGG19 to VGG16(the usage of VGG19 is also available, though). After a few tries, I got good results for some styles:

The wave:
![wave](https://cloud.githubusercontent.com/assets/7499023/21209728/0075796e-c2b1-11e6-853c-1564a8d426db.jpg)

The cubist:

![cubist](https://cloud.githubusercontent.com/assets/7499023/21209742/10087e9e-c2b1-11e6-9b9f-a63a7297a349.jpg)

I begin to think that it is the problem of the "style image" instead of the network. We got noisy results because the original image for "starry night" is very noisy:

![target_style_starry_tv](https://cloud.githubusercontent.com/assets/7499023/21209822/a0b19638-c2b1-11e6-8185-50963a659fc7.jpg)

I denoise the image and use this denoised image to retrain the model:

![target_style_denoised_starry_style250](https://cloud.githubusercontent.com/assets/7499023/21209831/b6182d52-c2b1-11e6-86f9-5f0af04faa36.jpg)

After denoising I got better results. It is the result for the original style image(not denoised):

![res](https://cloud.githubusercontent.com/assets/7499023/21209893/210922e2-c2b2-11e6-9ea7-61c799f01492.jpg)

Result for the denoised style image:

![res](https://cloud.githubusercontent.com/assets/7499023/21209942/6988388c-c2b2-11e6-94d4-ccf7bb2897f9.jpg)

The code is here: https://github.com/hzy46/fast-neural-style-tensorflow

**Comment by hzy46** [comment]
You can view my code: https://github.com/hzy46/fast-neural-style-tensorflow

**Comment by hzy46** [comment]
I forgot one thing.

In all my experiments, I set all tv_weight=0 and the tv_loss does not increase too much. The tv_loss always becomes steady after many iterations.

In the tensorboard:
![123](https://cloud.githubusercontent.com/assets/7499023/21215131/1cb7c4c8-c2da-11e6-829f-762faf8861cd.png)

Maybe we could get less noisy results by setting tv_weight. I have not experimented on this issue yet.

**Comment by OlavHN** [comment]
Thats awesome results!

Would be interesting to add some tv_loss to see if e.g. "the wave" transformation turned out even more smooth.

I'm really surprised by the jagged style picture. I took some measures to not distort it during preprocessing. The original sized style image is fetched here: [1] , then resized using bicubic interpolation, keeping the aspect ratio and finally center cropped here [2] before being fed to VGG.

Was the jagged style picture the result of the reader script resizing a high quality large style image?

A "bug" that got me for a long time was that previous versions of tensorboard (tensorflow visualisation dashboard) seemed to resize and badly distort the images.

Awesome job! I was getting nowhere :)

Also, thank you @jcjohnson for feedback and the raw images during debugging. Pretty awesome to get inputs from the horses mouth :)

[1] https://github.com/OlavHN/fast-neural-style/blob/master/fast_neural_style.py#L51
[2] https://github.com/OlavHN/fast-neural-style/blob/master/reader.py#L20

**Comment by dovanchan** [comment]
I saw you guys Intense debate， can you tell me which code is better that I can try to implement.Thanks!~

**Comment by OlavHN** [comment]
https://github.com/hzy46/fast-neural-style-tensorflow :)
